From 929de905ac525e36df83dd37ea2525162851f5fb Mon Sep 17 00:00:00 2001
From: Park Ju Hyung <qkrwngud825@gmail.com>
Date: Sun, 21 May 2017 16:05:33 +0900
Subject: [PATCH 047/296] PowerProfile: allow reading from a file instead of
 resource

Unified builds may require reading different power profile for each variants.

Allow reading power profile from a file instead of resource by querying
ro.power_profile prop.

Device maintainers can opt-in by using the init extension.
---
 .../java/com/android/internal/os/PowerProfile.java | 51 ++++++++++------------
 1 file changed, 24 insertions(+), 27 deletions(-)

diff --git a/core/java/com/android/internal/os/PowerProfile.java b/core/java/com/android/internal/os/PowerProfile.java
index 52718cbc276..314a61a4f53 100644
--- a/core/java/com/android/internal/os/PowerProfile.java
+++ b/core/java/com/android/internal/os/PowerProfile.java
@@ -21,13 +21,17 @@ import android.content.Context;
 import android.content.res.Resources;
 import android.content.res.XmlResourceParser;
 import android.os.SystemProperties;
-import android.util.Slog;
+import android.util.Xml;
 
 import com.android.internal.util.XmlUtils;
 
 import org.xmlpull.v1.XmlPullParser;
 import org.xmlpull.v1.XmlPullParserException;
 
+import libcore.io.IoUtils;
+
+import java.io.File;
+import java.io.FileReader;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -38,7 +42,6 @@ import java.util.HashMap;
  * [hidden]
  */
 public class PowerProfile {
-    private static final String TAG = "PowerProfile";
 
     /**
      * No power consumption, or accounted for elsewhere.
@@ -222,14 +225,27 @@ public class PowerProfile {
     }
 
     private void readPowerValuesFromXml(Context context) {
-        int id = getPowerProfileResId(context);
+        final String profilePath = SystemProperties.get("ro.power_profile",
+                                           "/system/etc/power_profile.xml");
         final Resources resources = context.getResources();
-        XmlResourceParser parser = resources.getXml(id);
+        XmlResourceParser resParser = null;
+        XmlPullParser parser = null;
         boolean parsingArray = false;
         ArrayList<Double> array = new ArrayList<Double>();
         String arrayName = null;
+        FileReader profileReader = null;
 
         try {
+            if (new File(profilePath).exists()) {
+                profileReader = new FileReader(profilePath);
+                parser = Xml.newPullParser();
+                parser.setInput(profileReader);
+            } else {
+                int id = com.android.internal.R.xml.power_profile;
+                resParser = resources.getXml(id);
+                parser = resParser;
+            }
+
             XmlUtils.beginDocument(parser, TAG_DEVICE);
 
             while (true) {
@@ -273,7 +289,10 @@ public class PowerProfile {
         } catch (IOException e) {
             throw new RuntimeException(e);
         } finally {
-            parser.close();
+            if (resParser != null) {
+                resParser.close();
+            }
+            IoUtils.closeQuietly(profileReader);
         }
 
         // Now collect other config variables.
@@ -376,28 +395,6 @@ public class PowerProfile {
         return 0;
     }
 
-    private int getPowerProfileResId(final Context context) {
-        int id = com.android.internal.R.xml.power_profile;
-        /*
-         * If ro.power_profile.override is set, use it to override the default.
-         * This is used for devices, which need to dynamically define the power profile.
-         */
-        String powerProfileOverride = SystemProperties.get("ro.power_profile.override");
-        if (!powerProfileOverride.isEmpty()) {
-            int tmpId = context.getResources().getIdentifier(powerProfileOverride, "xml",
-                    "android");
-            if (tmpId > 0) {
-                Slog.i(TAG, "getPowerProfileResId: using power profile \""
-                        + powerProfileOverride + "\"");
-                id = tmpId;
-            } else {
-                Slog.e(TAG, "getPowerProfileResId: could not retrieve power profile \""
-                        + powerProfileOverride + "\", using default instead");
-            }
-        }
-        return id;
-    }
-
     /**
      * Returns the number of memory bandwidth buckets defined in power_profile.xml, or a
      * default value if the subsystem has no recorded value.
-- 
2.11.0

